nmpolicy/internal/resolver/resolver.go

```diff
@@ -25,29 +25,60 @@ import (
 	"github.com/nmstate/nmpolicy/nmpolicy/types"
 )
 
-type Resolver struct {
-	currentState   map[string]interface{}
-	capturedStates map[string]map[string]interface{}
-	captureASTPool map[string]ast.Node
+type State struct {
+	CurrentState   map[string]interface{}
+	CapturedStates map[string]map[string]interface{}
+	CaptureASTPool map[string]ast.Node
 }
 
-func New() *Resolver {
-	return &Resolver{
-		currentState:   map[string]interface{}{},
-		capturedStates: map[string]map[string]interface{}{},
-		captureASTPool: map[string]ast.Node{},
+type ResolverWithState struct {
+	state State
+}
+
+type Resolver struct{}
+
+func NewState() State {
+	return State{
+		CurrentState:   map[string]interface{}{},
+		CapturedStates: map[string]map[string]interface{}{},
+		CaptureASTPool: map[string]ast.Node{},
+	}
+}
+
+func New() Resolver {
+	return Resolver{}
+}
+
+func NewWithState(state State) *ResolverWithState {
+	return &ResolverWithState{
+		state: state,
+	}
+}
+
+func (Resolver) Resolve(captureASTPool map[string]ast.Node, currentState []byte) (map[string]types.CaptureState, error) {
+	state := NewState()
+	state.CaptureASTPool = captureASTPool
+	return NewWithState(state).Resolve(currentState)
+}
+
+func (Resolver) ResolveCaptureEntryPath(captureEntryPathAST ast.Node, capturedStates map[string]types.CaptureState) (interface{}, error) {
+	var err error
+	state := NewState()
+	state.CapturedStates, err = unmarshalCapturedStates(capturedStates)
```

Comment:
To avoid multiple unmarshaling of the state I would move the unmarshal to the `CaptureEntry` creation `NewCaptureEntry`.